### PR TITLE
[Serve] Try enable SO_REUSEPORT beyond hasattr check

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This issue manifests when the host system do have port sharing but conda python binary was built against an older version of kernel which doesn't have this. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/http_proxy.py [hunk 1]
```diff
@@ -365,6 +365,19 @@ async def run(self):
...
+        # Therefore, in the absence of socket.SO_REUSEPORT, we try
+        # to use `15` which is value in linux kernel.
+        # https://github.com/torvalds/linux/blob/master/tools/include/uapi/asm-generic/socket.h#L27
+        else:
+            try:
+                sock.setsockopt(socket.SOL_SOCKET, 15, 1)
+            except Exception:
+                logger.debug(
+                    "Force setting SO_REUSEPORT failed. SO_REUSEPORT disabled."
+                )
```
Comment (edoakes) [inline python/ray/serve/http_proxy.py:380]: can we move this whole chunk of logic to a utility and test that it works as expected on the different platforms in CI?

Comment (edoakes) [reply python/ray/serve/http_proxy.py:380]: Let's also add an env var we can use as last resort to turn off REUSEPORT on all platforms

File: python/ray/serve/tests/test_standalone.py [hunk 2]
```diff
@@ -255,6 +275,9 @@ def third_actor_removed():
     wait_for_condition(third_actor_removed)
     ray.get(block_until_http_ready.remote("http://127.0.0.1:8005/-/routes"))
 
+    # TODO: remove before merge
+    assert False, "intentionally failing this to check linux CI"
```
Comment (edoakes) [inline python/ray/serve/tests/test_standalone.py:279]: don't forget this :)

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
ping @simon-mo

**Comment by simon-mo** [comment]
@shrekris-anyscale to take a closer look 🙏

File: python/ray/serve/http_proxy.py [hunk 3]
```diff
@@ -22,11 +23,15 @@
     RawASGIResponse,
     receive_http_body,
     Response,
+    set_socket_reuse_port,
 )
 from ray.serve.long_poll import LongPollClient
 
 MAX_REPLICA_FAILURE_RETRIES = 10
 DISCONNECT_ERROR_CODE = "disconnection"
+SOCKET_REUSE_PORT_ENABLED = (
```
Comment (shrekris-anyscale) [inline python/ray/serve/http_proxy.py:32]: Does this mean we disable port reuse by default from now on? Users have to explicitly set this env var to reuse ports.

File: python/ray/serve/tests/test_standalone.py (hunk 2 above)
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_standalone.py:279]: What were you checking in the CI?

Comment (simon-mo) [reply python/ray/serve/tests/test_standalone.py:279]: I'm checking that this tests get called in both macOS and Linux. Not just one of them.

File: python/ray/serve/http_proxy.py (hunk 3 above)
Comment (shrekris-anyscale) [reply python/ray/serve/http_proxy.py:32]: Never mind, I just noticed the default.

**Review by shrekris-anyscale — APPROVED** [review]

**Comment by simon-mo** [comment]
thanks @shrekris-anyscale for bring this to finish line

**Pull request merged**